### PR TITLE
add storage.erase_filesystem() to erase and reformat CIRCUITPY

### LINKED ISSUE #577: Consider adding a `storage.format()` to erase and create a fresh filesystem
It would be nice to be able to erase and re-create a damaged filesystem from the REPL or a program. Currently on non-Express boards we need to erase all of flash and then reload CircuitPython. On Express boards we have a special Arduino program that erases SPI flash, and then we need to reload CPy.

Something like `storage.format(<filesystem>)` would be helpful. One minor issue is how to specify the filesystem if it's so damaged we don't know what it's named.

Some previous discussion: https://github.com/adafruit/circuitpython/issues/386#issuecomment-356731896 ff.

### PR DESCRIPTION
**Pull request by dhalbert**
Fixes #577.

Note that `storage.erase_filesystem()` does a `microcontroller.reset()` after reformatting. This is necessary to get the host computer synced up with the filesystem, including both contents and mountpoint name.

The main use of this is to be called from the REPL, since if the filesystem is damaged, it's not likely one can write a file to call this anyway.

I purposely chose a long name to make it less likely to be used casually.

#577 discusses adding `storage.format(...)`, which could also be used on the SD card. That's a good idea, but a simple reformat of a filesystem visible to the host without syncing up leaves things in a bad state.

### CODE REVIEW

**Comment by ladyada** [comment]
this is rad!

**Comment by jerryneedell** [comment]
cool! worked great on a Trinket_m0!

**Comment by jerryneedell** [comment]
On itsybitsy_mo_express, it works, BUT - I have to press the reset button  after erasing the FS to get it to boot. After erasing the FS it tries to reboot, but hangs with a blue dotstar.

**Comment by jerryneedell** [comment]
hmm - tried itsybitsy again it it worked properly  (auto rebooted).

**Comment by jerryneedell** [comment]
Now it is back to failing regularly on my itsybitsy m0 espress.

**Comment by jerryneedell** [comment]
I am getting to following warning when compiling ```
jerryneedell@Ubuntu-Macmini:~/projects/adafruit_github/circuitpython_master/ports/atmel-samd$ make  BOARD=itsybitsy_m0_express 
Use make V=1, make V=2 or set BUILD_VERBOSE similarly in your environment to increase build verbosity.
install -d build-itsybitsy_m0_express/genhdr
python3 tools/gen_usb_descriptor.py \
	--manufacturer "Adafruit Industries LLC"\
	--product "Itsy Bitsy M0 Express"\
	--vid 0x239A\
	--pid 0x8012\
	--output_c_file build-itsybitsy_m0_express/autogen_usb_descriptor.c\
	--output_h_file build-itsybitsy_m0_express/genhdr/autogen_usb_descriptor.h
Generating build-itsybitsy_m0_express/genhdr/mpversion.h
GEN build-itsybitsy_m0_express/genhdr/qstr.i.last
QSTR updated
GEN build-itsybitsy_m0_express/genhdr/qstrdefs.generated.h
../../shared-bindings/usb_hid/__init__.h:33:23: warning: size of 'common_hal_usb_hid_devices' differ from the size of original declaration [-Wlto-type-mismatch]
 extern mp_obj_tuple_t common_hal_usb_hid_devices;
                       ^
common-hal/usb_hid/__init__.c:83:16: note: 'common_hal_usb_hid_devices' was previously declared here
 mp_obj_tuple_t common_hal_usb_hid_devices = {
                ^

51780 bytes free in flash out of 253440 bytes ( 247.5 kb ).
4152 bytes free in ram for stack out of 32768 bytes ( 32.0 kb ).

Create build-itsybitsy_m0_express/firmware.bin
Create build-itsybitsy_m0_express/firmware.uf2
python2 ../../tools/uf2/utils/uf2conv.py -b 0x2000 -c -o build-itsybitsy_m0_express/firmware.uf2 build-itsybitsy_m0_express/firmware.bin
Converting to uf2, output size: 403456, start address: 0x2000
Wrote 403456 bytes to build-itsybitsy_m0_express/firmware.uf2.
```

**Comment by jerryneedell** [comment]
Ah - Thanks for the reminder - I had forgotten about this. I think I am going to get a lot of mileage out of this PR!

**Comment by dhalbert** [comment]
@jerryn You mean the USB port becomes unusable? The reset done by `erase_filesystem()` is basically the same as pulling the plug or pressing the reset button. When I tried it on Ubuntu it just remounted CIRCUITPY, and I was able to get into the REPL again. Is this on Raspbian?

**Comment by jerryneedell** [comment]
This is on Ubuntu 16.04

**Comment by jerryneedell** [comment]
just tried again - it erases the FS but CIRCUITPY is not unmounted - it still shows the original FS - If I unmount I cannot recount it. reseting the M4 result in the error above. Rebooting the Ubuntu machine brings it all back.

**Comment by jerryneedell** [comment]
Tried it on a metro_m0_express - worked fine.

**Comment by jerryneedell** [comment]
BTW - the breaks the esp8266 ```
FREEZE modules
build/shared-module/storage/__init__.o:(.text.common_hal_storage_erase_filesystem+0x0): undefined reference to `filesystem_init'
build/shared-module/storage/__init__.o: In function `common_hal_storage_erase_filesystem':
__init__.c:(.text.common_hal_storage_erase_filesystem+0xd): undefined reference to `filesystem_init'
Makefile:249: recipe for target 'build/firmware.elf' failed
make: *** [build/firmware.elf] Error 1
```

File: ports/atmel-samd/supervisor/filesystem.c [hunk 1]
```diff
@@ -50,7 +50,7 @@ static void make_empty_file(FATFS *fatfs, const char *path) {
 
 // we don't make this function static because it needs a lot of stack and we
 // want it to be executed without using stack within main() function
-void filesystem_init(bool create_allowed) {
+void filesystem_init(bool create_allowed, bool create_force) {
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/atmel-samd/supervisor/filesystem.c]: sure

**Comment by dhalbert** [comment]
@jerryneedell I am also using Ubuntu 16.04 x64, and don't see this with the M4 on my machine (a vanilla Dell Optiplex, 2013 or so). So it may depend on the low-level USB drivers or might be a timing problem.

I don't necessarily think this will help, but could you try adding a 1-second delay before resetting, in `ports/atmel_samd/shared-module/storage/__Init__.c` ?

At the top:
```
#include "py/mphal.h"
```
at the end of the file, add a delay call:
```
void common_hal_storage_erase_filesystem(void) {
    filesystem_init(false, true);   // Force a re-format.
    mp_hal_delay_ms(1000);
    common_hal_mcu_reset();
    // We won't actually get here, since we're resetting.
}
```
If that doesn't help, could you try a USB3 vs USB2 port, and/or a USB hub?

How are you mounting CIRCUITPY when it reboots? I am just clicking CIRCUITPY in a Nautilus windows (or the equivalent); sometimes I don't even need to do that: it's already mounted. I don't use command line `mount`.

**Comment by jerryneedell** [comment]
ading the dealy did not have any impact, but connecting the M4 to a USB port directly on the Host machine (no hub)  does work.

**Comment by jerryneedell** [comment]
Tried on M4 revb on MacOS  -- remounts successfully - even on USB2 Hub
Tried on Raspian to Pi 3B+  -  fails to unmount just as Ubuntu - power cycling hub remounts it successfully.

**Comment by dhalbert** [comment]
On ESP8266: throw `NotImplementedError`. It would be possible to erase the filesystem but then the needed `boot.py` would not be present. It's easier to just erase flash and re-upload CircuitPython. The `NotImplementedError` suggests this.

On nrf: placeholder `NotImplementedError` for now.

**Comment by tannewt** [comment]
Raising NotImplementedError makes sense to me because the ESP8266 is much less likely to have corruption since it doesn't do USB mass storage.

**Pull request merged**